packages/eslint-plugin/src/rules/strict-boolean-expressions.ts

```diff
@@ -232,6 +242,111 @@ export default createRule<Options, MessageId>({
       | TSESTree.IfStatement
       | TSESTree.WhileStatement;
 
+    function isBooleanType(expressionType: ts.Type): boolean {
+      return tsutils.isTypeFlagSet(
+        expressionType,
+        ts.TypeFlags.Boolean | ts.TypeFlags.BooleanLiteral,
+      );
+    }
+
+    function isFunctionReturningBooleanType(type: ts.Type): boolean {
+      for (const signature of type.getCallSignatures()) {
+        const returnType = signature.getReturnType();
+
+        if (tsutils.unionTypeParts(returnType).every(isBooleanType)) {
+          return true;
+        }
+      }
+
+      return false;
+    }
+
+    function isTypeParameter(type: ts.Type): boolean {
+      // https://github.com/JoshuaKGoldberg/ts-api-utils/issues/382
+      return (tsutils.isTypeParameter as (type: ts.Type) => boolean)(type);
+    }
+
+    function isParameterAssignableTo(
+      parameter: ts.Symbol,
+      expected: ts.Type,
+    ): boolean {
+      const parameterType = checker.getTypeOfSymbol(parameter);
+
+      const constrained = isTypeParameter(parameterType)
+        ? checker.getBaseConstraintOfType(parameterType)
+        : parameterType;
+
+      // treat a type parameter without constraint as `unknown`
+      if (!constrained) {
+        return true;
+      }
```

Comment:
[This comment](https://github.com/typescript-eslint/typescript-eslint/pull/10314#discussion_r1836097699) is probably relevant here as well.